### PR TITLE
Fixed #24649 -- Allowed using Avg aggregate on non-numeric field types.

### PR DESCRIPTION
**Pull request by timgraham**
https://code.djangoproject.com/ticket/24649

### CODE REVIEW

File: docs/ref/models/querysets.txt [hunk 1]
```diff
@@ -2802,12 +2802,19 @@ by the aggregate.
...
+    unless you specify a different ``output_field``.
 
     * Default alias: ``<field>__avg``
-    * Return type: ``float``
+    * Return type: ``float`` (or the type of whatever ``output_field`` is
+      specified)
+
+    .. versionchanged:: 1.9
+
+        The ``output_field`` parameter was added to allow aggregating across
```
Comment (jarshwah) [inline docs/ref/models/querysets.txt]: "aggregating over" rather than "aggregating across" ? Same below.

**Comment by jarshwah** [comment]
I would like to see a test that does `Avg('decimal_field')` to ensure that floats are still being returned from all backends with the removal of Avg.convert_value. Otherwise good patch.

**Comment by timgraham** [comment]
Updated per comments.

The new `DurationField` test doesn't work on Oracle: "DatabaseError: ORA-00932: inconsistent datatypes: expected NUMBER got INTERVAL DAY TO SECOND". Can you advise there?

**Comment by jarshwah** [comment]
Hmm Oracle doesn't allow Avg or Sum on interval data types. Seems like the commonly accepted way to do so is with `EXTRACT(day from 24*60*60*interval)`. See http://stackoverflow.com/a/5063553/10583. That only gets you to second resolution though, completely ignoring microseconds. Not really ideal.

http://www.dba-oracle.com/t_timestamp_math_elapsed_times.htm is another option which involves extracting each piece of the date component and multiplying by a constant. That'll give microsecond precision (decimal second component at least).

```
with the_interval as (
  select INTERVAL '4 5:12:10.999999' DAY TO SECOND(6) itv from dual
)
select 
  itv,
 -- convert back to interval after applying avg function
 NUMTODSINTERVAL(avg(extract(day from itv)*86400 +
  extract(hour from itv)*3600 +
  extract(minute from itv)*60 +
  extract(second from itv)), 'SECOND') seconds_with_precision,
  extract(day from 24*60*60*itv) seconds_without_precision
from the_interval;
```

To support this in AVG, we'd need to create an as_oracle method, test if the output_field is an interval, and then wrap output_field with multiple extracts. That's getting into pretty gnarly territory though. The functions.Coalesce does something similar with TextFields, so it's not unprecedented.

If you don't want to deal with this (and I wouldn't blame you), then if you leave the ticket open with this detail linked, I'll come back to it when I've got a bit of spare time.

**Comment by timgraham** [comment]
buildbot, test on Oracle.

**Comment by timgraham** [comment]
buildbot, test on oracle.

**Comment by timgraham** [comment]
Added a database feature to skip the test on Oracle for now and created #24699 to implement Oracle support.

**Pull request merged**